python/pyiceberg/table/scan.py

```diff
@@ -0,0 +1,115 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import itertools
+from abc import ABC, abstractmethod
+from typing import Iterable
+
+from pyiceberg.expressions import AlwaysTrue, BooleanExpression
+from pyiceberg.expressions.visitors import manifest_evaluator
+from pyiceberg.io import FileIO
+from pyiceberg.manifest import DataFile, ManifestFile
+from pyiceberg.table import PartitionSpec, Snapshot, Table
+
+
+class FileScanTask(ABC):
+    """A scan task over a range of bytes in a single data file."""
+
+    manifest: ManifestFile
+    data_file: DataFile
+    residual: BooleanExpression | None
+    spec: PartitionSpec | None
+
+    def __init__(
+        self,
+        manifest: ManifestFile,
+        data_file: DataFile,
+        spec: PartitionSpec | None = None,
+        residual: BooleanExpression | None = None,
+    ):
+        self.manifest = manifest
+        self.data_file = data_file
+        self.spec = spec
+        self.residual = residual
+
+    @property
+    def start(self) -> int:
+        return 0
+
+    @property
+    def length(self) -> int:
+        return self.data_file.file_size_in_bytes
+
+
+class TableScan(ABC):
+    """API for configuring a table scan."""
+
+    table: Table
+    snapshot: Snapshot
+    expression: BooleanExpression
+
+    def __init__(self, table: Table, snapshot: Snapshot | None = None, expression: BooleanExpression | None = None):
+        self.table = table
+        self.expression = AlwaysTrue() if expression is None else expression
+        if snapshot is None:
+            snapshot = table.current_snapshot()
+        if snapshot is not None:
+            self.snapshot = snapshot
+        else:
+            raise ValueError("Unable to resolve to a Snapshot to use for the table scan.")
+
+    @abstractmethod
+    def plan_files(self) -> Iterable[FileScanTask]:
+        """Plan tasks for this scan where each task reads a single file.
+
+        Returns:
+            Table: a tuple of tasks scanning entire files required by this scan
+        """
+
+
+class DataTableScan(TableScan):
+    """API for configuring a table scan."""
+
+    io: FileIO
+
+    def __init__(self, io: FileIO, table: Table, snapshot: Snapshot | None = None, expression: BooleanExpression | None = None):
```

Comment:
How about passing in the schema and partition-spec, rather than pulling in the whole table object?